cap-primitives/src/std/fs/canonicalize.rs

```diff
@@ -28,7 +29,36 @@ pub fn canonicalize(start: &fs::File, path: &Path) -> io::Result<PathBuf> {
             (Err(path_err), Err(canonical_err)) => {
                 assert_eq!(path_err.to_string(), canonical_err.to_string())
             }
-            other => panic!("Inconsistent canonicalize checks: {:?}", other),
+            other => panic!("inconsistent canonicalize checks: {:?}", other),
+        }
+
+        // On operating systems which can tell us the path of a file descriptor,
+        // assert that the path we computed canonicalizes to the same thing as
+        // the input canonicalizes too.
+        if let Some(start_abspath) = get_path(start) {
+            let mut check_abspath = start_abspath.clone();
+            let mut result_abspath = start_abspath.clone();
+            check_abspath.push(path);
+            result_abspath.push(canonical_path);
+            let check_abspath =
+                fs::canonicalize(check_abspath).expect("we already canonicalized this before");
+            let result_abspath =
+                fs::canonicalize(result_abspath).expect("we already canonicalized this before");
+            assert_eq!(
+                check_abspath,
+                result_abspath,
+                "incorrect canonicalization: start='{:?}' path='{}' result='{}'",
+                start,
+                path.display(),
+                canonical_path.display()
+            );
+            assert!(
+                result_abspath.starts_with(start_abspath),
```

Comment:
I think this check might fail on Windows since `start_abspath` will *not* contain the extended path prefix `\\?\`, while the result of `fs::canonicalize` always does:

> On Windows, this converts the path to use extended length path syntax...

Anyhow, I reckon this is OK to leave as-is for now since we don't have a working implementation on Windows yet anyhow, but if you could leave a comment to that effect as a reminder, that'd be great!

Reply:
Commented!